seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/FileMaterializers.scala

```diff
@@ -1,41 +1,76 @@
 package org.scalajs.jsenv.selenium
 
+import scala.collection.JavaConverters._
+
 import java.io._
 import java.nio.file._
 import java.net._
 
-import org.scalajs.core.tools.io._
-import org.scalajs.jsenv.VirtualFileMaterializer
+import org.scalajs.io._
 
-private[selenium] trait FileMaterializer {
-  def materialize(vf: VirtualTextFile): URL
-}
+private[selenium] abstract class FileMaterializer(bypassRealFiles: Boolean) {
+  private val tmpSuffixRE = """[a-zA-Z0-9-_.]*$""".r
+
+  private[this] var tmpFiles: List[Path] = Nil
+
+  final def materialize(vf: VirtualBinaryFile): URL = vf match {
+    case vf: FileVirtualBinaryFile if bypassRealFiles => vf.file.toURI.toURL
+
+    case vf =>
+      val tmp = newTmp(vf.path)
+      val in = vf.inputStream
+      Files.copy(in, tmp, StandardCopyOption.REPLACE_EXISTING)
+      toURL(tmp)
+  }
 
-/** Materializes virtual files in a temp directory (uses file:// schema). */
-private[selenium] object TempDirFileMaterializer extends FileMaterializer {
-  private val materializer = new VirtualFileMaterializer(singleDir = false)
+  final def materialize(name: String, content: String): URL = {
+    val tmp = newTmp(name)
+    Files.write(tmp, Iterable(content).asJava)
+    toURL(tmp)
+  }
+
+  final def close(): Unit = tmpFiles.foreach(Files.delete)
 
-  override def materialize(vf: VirtualTextFile): URL = {
-    materializer.materialize(vf).toURI.toURL
+  private def newTmp(path: String): Path = {
+    val suffix = tmpSuffixRE.findFirstIn(path).orNull
+    val p = createTmp(suffix)
+    tmpFiles ::= p
+    p
   }
+
+  protected def createTmp(suffix: String): Path
+  protected def toURL(file: Path): URL
 }
 
-private[selenium] class ServerDirFileMaterializer(contentDir: Path,
-    webRoot: URL) extends FileMaterializer {
+object FileMaterializer {
+  import SeleniumJSEnv.Config.Materialization
+  def apply(m: Materialization): FileMaterializer = m match {
+    case Materialization.Temp =>
+      new TempDirFileMaterializer
 
-  def this(contentDir: String, webRoot: String) =
-    this(Paths.get(contentDir), new URL(webRoot))
+    case Materialization.Server(contentDir, webRoot) =>
+      new ServerDirFileMaterializer(contentDir, webRoot)
+  }
+}
 
-  require(webRoot.getPath().endsWith("/"), "webRoot must end with a slash (/)")
```

Comment:
It is in `Config.Materialization.Server`. Is that good enough or do you want me to duplicate it?

Reply:
Ah, I see. It wasn't part of the diff, which is why I didn't see it. It's good enough, since we can only call `FileMaterialization.apply` with an existing `Config.Materialization`, so the `require` must have been executed.